crates/sui-graphql-rpc/src/server/builder.rs

```diff
@@ -237,29 +250,54 @@ impl ServerBuilder {
         }
 
         // TODO: uncomment once impl
-        // if config.internal_features.open_telemetry {
-        //     let tracer;
-        //     builder = builder.extension(OpenTelemetry::new(tracer));
-        // }
+        // if config.internal_features.open_telemetry { }
 
         Ok(builder)
     }
 }
 
 async fn graphql_handler(
+    State(metrics): State<Option<Metrics>>,
     ConnectInfo(addr): ConnectInfo<SocketAddr>,
     schema: axum::Extension<SuiGraphQLSchema>,
     headers: HeaderMap,
     req: GraphQLRequest,
 ) -> GraphQLResponse {
+    if let Some(ref m) = metrics {
+        m.request_metrics.inflight_requests.inc();
+    }
+    let instant = Instant::now();
     let mut req = req.into_inner();
+    req.data.insert(Uuid::new_v4());
     if headers.contains_key(ShowUsage::name()) {
         req.data.insert(ShowUsage)
     }
     // Capture the IP address of the client
     // Note: if a load balancer is used it must be configured to forward the client IP address
     req.data.insert(addr);
-    schema.execute(req).await.into()
+    let result = schema.execute(req).await;
+    let elapsed = instant.elapsed().as_millis() as u64;
+    if let Some(m) = metrics {
+        m.query_latency(elapsed);
+        m.inc_num_queries();
+        if !result.is_ok() {
+            m.inc_errors(result.errors.clone());
+        }
+        m.request_metrics.inflight_requests.dec();
+
+        // TODO Is this the right way to get the top level query path names?
```

Comment:
That sounds better. I wonder if there is an overhead for using this `resolve` hook instead of doing it once everything got parsed and returned. I added the hook and we can do some experiments later.

Reply:
```graphql
{
  epoch1: epoch {
    epochId
  }
  epoch2: epoch {
    epochId 
  }
  chainIdentifier
}
```
results in 
```
# HELP num_queries_top_level Number of queries for each top level node
# TYPE num_queries_top_level counter
num_queries_top_level{path="chainIdentifier"} 1
num_queries_top_level{path="epoch"} 2
```